File: mobile/android/app/src/main/java/fyi/dayle/data/MissionService.kt
package fyi.dayle.data

import org.json.JSONArray
import org.json.JSONObject
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets

class MissionService(private val baseUrl: String) {

    fun fetchMission(previousMissions: List<String>): String {
        val connection = (URL("$baseUrl/mission").openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            connectTimeout = 15_000
            readTimeout = 30_000
            doOutput = true
            setRequestProperty("Content-Type", "application/json; charset=UTF-8")
        }

        try {
            val body = JSONObject().apply {
                put("previousMissions", JSONArray(previousMissions))
            }.toString()

            OutputStreamWriter(connection.outputStream, StandardCharsets.UTF_8).use {
                it.write(body)
                it.flush()
            }

            val code = connection.responseCode
            val stream = if (code in 200..299) connection.inputStream else connection.errorStream
            val responseText =
                stream?.bufferedReader(StandardCharsets.UTF_8)?.use { it.readText() }.orEmpty()

            if (code !in 200..299) {
                throw MissionServiceException("HTTP $code: $responseText")
            }

            val parsed = JSONObject(responseText)
            val mission = parsed.optString("mission").trim()
            if (mission.isEmpty()) {
                throw MissionServiceException("Empty mission in response")
            }
            return mission
        } finally {
            connection.disconnect()
        }
    }
}

class MissionServiceException(message: String) : Exception(message)

File: app/lib/services/mission_service.dart
import "dart:convert";

import "package:http/http.dart" as http;

class MissionService {
  static const _baseUrl = "https://dayle.francids.workers.dev";

  Future<String> fetchMission(List<String> previousMissions) async {
    final uri = Uri.parse("$_baseUrl/mission");
    final body = jsonEncode({"previousMissions": previousMissions});

    final response = await http
        .post(
          uri,
          headers: {"Content-Type": "application/json; charset=UTF-8"},
          body: body,
        )
        .timeout(const Duration(seconds: 30));

    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw MissionServiceException(
        "HTTP ${response.statusCode}: ${response.body}",
      );
    }

    final parsed = jsonDecode(response.body) as Map<String, dynamic>;
    final mission = (parsed["mission"] as String).trim();
    if (mission.isEmpty) {
      throw MissionServiceException("Empty mission in response");
    }
    return mission;
  }
}

class MissionServiceException implements Exception {
  final String message;
  MissionServiceException(this.message);

  @override
  String toString() => message;
}
